README.md
## Allgemeines

> Beim erstellen einer monatlichen Transaktion ist zu beachten dass, nicht geprüft wird ob es sich um den letzten Tag des Monats handelt.
> D.h. sollte eine Transaktion am 29 Februar starten würde diese am 29. April ausgeführt werden und nicht am 30. April
> Daher ist zu empfehlen eine solche Transaktion auf den jewails ersten Tag des Monats auszuführen

```
PRODUCTION=true
APPLICATION=<ENTER>

SUPABASE_URL=<ENTER>
SUPABASE_ANON_KEY=<ENTER>

DB_HOST=<ENTER>
DB_USER=<ENTER>
DB_PASSWORD=<ENTER>
DB_DATABASE=<ENTER>

TRANSFER_TOKEN=<ENTER>
LARAVEL=<ENTER>
XSRF=<ENTER>
```

## Datenbank

| Spalte      | Typ                               |
| ----------- | --------------------------------- |
| `id`        | `int(10) unsigned auto-increment` |
| `execution` | `int(10) unsigned`                |
| `sender`    | `varchar(8)`                      |
| `receiver`  | `varchar(8)`                      |
| `balance`   | `double`                          |
| `note`      | `varchar(80)`                     |

| Spalte       | Typ                                  |
| ------------ | ------------------------------------ |
| `id`         | `int(10) unsigned auto-increment`    |
| `type`       | `DAILY / WEEKLY / MONTHLY / PLANNED` |
| `start_date` | `date`                               |
| `end_date`   | `date`                               |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.6](https://github.com/DulliAG/automated-transactions/compare/v1.1.5...v1.1.6) (2022-07-23)


src/tests/validate_iban.test.ts
import { TransactionService } from '../service/transaction.service';

describe('Check if provided IBAN is valid', () => {
  test('(Valid) NH092843', () => {
    expect(TransactionService.validateIban('NH092843')).toBe(true);
  });

  test('(Invalid) NH09e84e', () => {
    expect(TransactionService.validateIban('NH09e84e')).toBe(false);
  });

  test('(Invalid) NH09284f', () => {
    expect(TransactionService.validateIban('NH09284f')).toBe(false);
  });

  test('(Invalid) WH092843', () => {
    expect(TransactionService.validateIban('WH092843')).toBe(false);
  });
});


src/tests/should_execute.test.ts
import { TransactionService } from '../service/transaction.service';

describe('Check if the transaction should execute', () => {
  test('Daily, but start-date is in the future', () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'DAILY',
          start_date: '2022-07-05',
          end_date: '2022-07-14',
        },
        new Date('2022-07-03')
      )
    ).toBe(false);
  });

  test('Daily, should execute during period', () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'DAILY',
          start_date: '2022-07-03',
          end_date: '2022-07-14',
        },
        new Date('2022-07-03')
      )
    ).toBe(true);
  });

  test("Daily, should'nt execute after period", () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'DAILY',
          start_date: '2022-07-01',
          end_date: '2022-07-02',
        },
        new Date('2022-07-03')
      )
    ).toBe(false);
  });

  test('Weekly, should execute on start-date', () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'WEEKLY',
          start_date: '2022-07-03',
          end_date: '2022-07-30',
        },
        new Date('2022-07-03')
      )
    ).toBe(true);
  });

  test('Weekly, should execute one week after start', () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'WEEKLY',
          start_date: '2022-07-03',
          end_date: '2022-07-30',
        },
        new Date('2022-07-10')
      )
    ).toBe(true);
  });

  test("Weekly, should'nt execute", () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'WEEKLY',
          start_date: '2022-07-03',
          end_date: '2022-07-30',
        },
        new Date('2022-07-11')
      )
    ).toBe(false);
  });

  test('Month, run on start-date', () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'MONTHLY',
          start_date: '2022-02-28',
          end_date: '2022-07-30',
        },
        new Date('2022-02-28')
      )
    ).toBe(true);
  });

  test("Month, doesn't contains date", () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'MONTHLY',
          start_date: '2022-02-29',
          end_date: '2022-07-30',
        },
        new Date('2022-03-31')
      )
    ).toBe(false);
  });

  test('Month, run 2 months after start', () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'MONTHLY',
          start_date: '2022-02-14',
          end_date: '2022-05-30',
        },
        new Date('2022-03-14')
      )
    ).toBe(true);
  });

  test('Month, run 3 months after start on last day', () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'MONTHLY',
          start_date: '2022-02-28',
          end_date: '2022-06-30',
        },
        new Date('2022-05-31')
      )
    ).toBe(true);
  });

  test('Planned, should execute on start and end-date', () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'PLANNED',
          start_date: '2022-07-03',
          end_date: '2022-07-03',
        },
        new Date('2022-07-03')
      )
    ).toBe(true);
  });

  test("Planned, should'nt execute after start and end-date", () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'PLANNED',
          start_date: '2022-07-03',
          end_date: '2022-07-03',
        },
        new Date('2022-07-04')
      )
    ).toBe(false);
  });

  test("Planned, should'nt execute on different start and end-date", () => {
    expect(
      TransactionService.validateExecutionDate(
        {
          id: 1,
          type: 'PLANNED',
          start_date: '2022-07-03',
          end_date: '2022-07-04',
        },
        new Date('2022-07-04')
      )
    ).toBe(false);
  });
});


src/index.ts
import dotenv from 'dotenv';
dotenv.config();
import { CronJob } from 'cron';
import { format } from 'date-fns';

import { client } from './service/log.service';
import { supabase } from './supabase';
import { Transaction } from './interface/transaction.interface';
import { TransactionService } from './service/transaction.service';
import { PRODUCTION } from './constants';

const TRANSACTION_TIMEOUT = 500;

client.log('LOG', 'Setup', `Starting ${process.env.APPLICATION}...`);
const processTransactions = new CronJob('0 1 * * *', async () => {
  try {
    client.log('LOG', 'Retrieve transactions', 'Rufe Daten aus der Datenbank ab');
    const { data, error } = await supabase
      .from<Transaction>('bkr_transaction')
      .select(
        `
          id,
          sender,
          receiver,
          balance, 
          note,
          bkr_execution!inner (
            id,
            type,
            start_date,
            end_date
          ) 
        `
      )
      // If we're processing an planned execution the start_date and the end_date are equal and represent the execution_date
      .lte(
        // @ts-expect-error
        'bkr_execution.start_date',
        format(new Date(), 'yyyy-MM-dd')
      )
      .gte(
        // @ts-expect-error
        'bkr_execution.end_date',
        format(new Date(), 'yyyy-MM-dd')
      );
    if (error) throw error;

    const transactionsForExecution = data.filter((transaction) =>
      TransactionService.validate(
        transaction.sender,
        { target: transaction.receiver, amount: transaction.balance, info: transaction.note },
        transaction.bkr_execution
      )
    );
    await client.log(
      'LOG',
      'Process transactions',
      JSON.stringify({
        message: 'Auswerten der gültigen Transaktionen',
        transactions: transactionsForExecution,
      })
    );

    // Execute transactions
    const transactionService = new TransactionService();
    transactionsForExecution.forEach((transaction, index) => {
      setTimeout(async () => {
        let options = {
          target: transaction.receiver,
          amount: transaction.balance,
          info: transaction.note,
        };
        if (!PRODUCTION) options.amount = 1; // Because we don't wanna spent or loose all our money we're only gonna transfer 1 $ per transaction
        return transactionService
          .transfer(transaction.sender, options)
          .then((result) => {
            client.log(
              'LOG',
              'Transfer money',
              JSON.stringify({ id: transaction.id, details: options, message: result })
            );
          })
          .catch((err) => client.log('ERROR', 'Transfer money', err));
      }, index * TRANSACTION_TIMEOUT);
    });
  } catch (error) {
    // @ts-expect-error
    client.log('ERROR', 'Uncategorized', error);
  }
});

if (!PRODUCTION) processTransactions.fireOnTick();
processTransactions.start();


src/service/transaction.service.ts
import dotenv from 'dotenv';
dotenv.config();
import axios from 'axios';
import { isSameDay, differenceInDays, addHours } from 'date-fns';
import { TransactionDetails, TransactionExecution } from '../interface/transaction.interface';

export class TransactionService {
  private COOKIES = {
    XSRF: `XSRF-TOKEN=${process.env.XSRF!}`,
    LARAVEL: `laravel_session=${process.env.LARAVEL!}`,
  };

  private OPTIONS = {
    headers: {
      cookie: `${this.COOKIES.XSRF} ${this.COOKIES.LARAVEL}`,
    },
  };

  private TOKEN = process.env.TRANSFER_TOKEN!;

  static validateIban(iban: string) {
    return new RegExp('NH[0-9]{6}').test(iban) && iban.length === 8;
  }

  static validateExecutionDate(
    { type, start_date, end_date }: TransactionExecution,
    now: Date = addHours(new Date(), 2)
  ) {
    const startDate = new Date(start_date),
      endDate = new Date(end_date);

    const isInDateRange = now >= startDate && now <= endDate;

    const isStartDay = isSameDay(now, startDate);

    switch (type) {
      case 'DAILY':
        return isInDateRange;

      case 'WEEKLY':
        return isInDateRange && (isStartDay || differenceInDays(now, startDate) % 7 === 0);

      case 'MONTHLY':
        // Check if the start-date is the last day of the month
        // If this is the case we're gonna execute the transaction on the last day of the following months
        // If this is not the case the transactions is gonna get executed next month on the same date
        const getLastDateOfMonth = (year: number, month: number) =>
          new Date(year, month, 0).getDate();

        /**
         * Prüfe ob
         * - prüfe ob start-datum letzter tag des monats ist
         *   (wenn dies der fall ist soll die zahlung immer am letzten tag der folgemonate ausgeführt werden)
         */

        const lastDayOfStartMonth = getLastDateOfMonth(
          Number(start_date.split('-')[0]),
          Number(start_date.split('-')[1])
        );
        const startDateIsLastDateOfMonth = lastDayOfStartMonth === Number(start_date.split('-')[2]);
        const todayIsLastDateOfMonth =
          getLastDateOfMonth(now.getFullYear(), now.getDate()) === now.getDate();

        // e.g. 2022-02-29 would be false because the last day would be 2022-02-28
        const startMonthContainsDate =
          lastDayOfStartMonth > 0 && Number(start_date.split('-')[2]) <= lastDayOfStartMonth;
        const monthHasPassed = startDate.getDate() === now.getDate();

        return (
          isInDateRange &&
          startMonthContainsDate &&
          (isStartDay ||
            (startDateIsLastDateOfMonth && todayIsLastDateOfMonth) ||
            (!isStartDay && monthHasPassed))
        );

      case 'PLANNED':
        // If we're checking an PLANNED-execution the start- and end-date are the same.
        // If this is not the case were checking an defective transaction
        return isInDateRange && isStartDay && isSameDay(now, endDate);
    }
  }

  static validate(
    sender: string,
    { target, amount, info }: TransactionDetails,
    exeuction: TransactionExecution
  ) {
    if (amount < 0) {
      throw new Error(
        JSON.stringify({
          message: `Transfer amount has to be 1 or greater`,
          error: '*/invalid-amount',
        })
      );
    }

    if (sender === target) {
      throw new Error(
        JSON.stringify({
          message: `The reciver is identical with the sender`,
          error: '*/equal-receiver',
        })
      );
    }

    if (!this.validateIban(sender)) {
      throw new Error(
        JSON.stringify({ message: `Provided 'sender' IBAN is invalid`, error: '*/invalid-iban' })
      );
    }

    if (!this.validateIban(target)) {
      throw new Error(
        JSON.stringify({ message: `Provided 'target' IBAN is invalid`, error: '*/invalid-iban' })
      );
    }

    return this.validateExecutionDate(exeuction);
  }

  getTransactions(iban: string, options = this.OPTIONS) {
    return new Promise((res, rej) => {
      axios
        .post(`https://info.realliferpg.de/banking/${iban}/data`, {}, options)
        .then((response) => res(response.data))
        .catch((err) => rej(err));
    });
  }

  transfer(
    iban: string,
    { target, amount, info }: TransactionDetails,
    token = this.TOKEN,
    options = this.OPTIONS
  ): Promise<string> {
    return new Promise((res, rej) => {
      axios
        .post(
          `https://info.realliferpg.de/banking/${iban}`,
          {
            _token: token,
            type: 'init_transaction',
            amount: amount.toString(),
            iban: target,
            info: info,
          },
          options
        )
        .then((response) => {
          // if (!RESULT.includes('Deine Überweisung wurde aufgegeben und durchgeführt!')) {
          //   if (RESULT.includes('Falsches IBAN-Format!')) throw new Error('Falsches IBAN-Format!');

          //   if (RESULT.includes('Zu wenig Geld auf dem Konto!'))
          //     throw new Error('Zu wenig Geld auf dem Konto!');

          //   if (RESULT.includes('Zielkonto existiert nicht'))
          //     throw new Error('Zielkonto existiert nicht!');

          //   // console.log(RESULT);
          //   throw new Error('Ein unbekannter Fehler ist aufgetreten!');
          // }

          res('Das Geld wurde überwiesen');
        })
        .catch((err) => rej(err));
    });
  }
}
